bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Shell.java

```diff
@@ -298,8 +300,24 @@ public Shell (Display display, int style) {
 	if (handle != 0 && !embedded) {
 		state |= FOREIGN_HANDLE;
 	}
+
+	int shellZoom;
+	int shellNativeZoom;
+	if (parent != null) {
+		shellZoom = parent.getZoom();
+		shellNativeZoom = parent.getNativeZoom();
+	} else {
+		int mappedDPIZoom = getMonitor().getZoom();
+		shellZoom = DPIUtil.getZoomForAutoscaleProperty(mappedDPIZoom);
+		shellNativeZoom = mappedDPIZoom;
+	}
+	this.setZoom(shellZoom);
+	this.setNativeZoom(shellNativeZoom);
+
 	reskinWidget();
 	createWidget ();
+
+	addListener(SWT.ZoomChanged, this::handleZoomEvent);
```

Comment:
nit: wouldn't it be better to add the listener only when `autoScaleOnRuntime` is true instead of checking that in the `handleZoomEvent` method?

Reply:
Good point